# Bank Customer Data Preprocessing

## Importing python modules and data

Import modules

In [1]:
# Import modules
import zipfile
import pandas as pd
import numpy as np
import matplotlib as plt
import sys

Import data

In [2]:
# Import data - Zip file
path_to_zip_file = r'D:\Project\Bank customer segmentation\customer_segmentation\data\raw\bank_transactions.csv.zip'
with zipfile.ZipFile(path_to_zip_file, 'r') as zip_ref:
    zip_ref.extractall(r'D:\Project\Bank customer segmentation\customer_segmentation\data\raw')

In [2]:
# Import zip file
filepath = r'D:\Project\Bank customer segmentation\customer_segmentation\data\raw\bank_transactions.csv'
with open(filepath,'r'):
    df = pd.read_csv(filepath)

# View file import
df.head()

TransactionID CustomerID CustomerDOB CustGender CustLocation  \
0            T1   C5841053     10/1/94          F   JAMSHEDPUR   
1            T2   C2142763      4/4/57          M      JHAJJAR   
2            T3   C4417068    26/11/96          F       MUMBAI   
3            T4   C5342380     14/9/73          F       MUMBAI   
4            T5   C9031234     24/3/88          F  NAVI MUMBAI   

   CustAccountBalance TransactionDate  TransactionTime  \
0            17819.05          2/8/16           143207   
1             2270.69          2/8/16           141858   
2            17874.44          2/8/16           142712   
3           866503.21          2/8/16           142714   
4             6714.43          2/8/16           181156   

   TransactionAmount (INR)  
0                     25.0  
1                  27999.0  
2                    459.0  
3                   2060.0  
4                   1762.5

## Dataset information

In [6]:
# Create output file to store results
# outfile = r'D:\Project\Bank customer segmentation\customer_segmentation\reports\1_outfile_eda.txt'
# Shape of dataset
nrows, ncols = df.shape
print('No of rows: ', nrows,'\n','No of columns: ', ncols)
# Dataset info
print(df.info())

# Dataset columns
column_names = df.columns
print(df.columns)

No of rows:  1048567 
 No of columns:  9
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048567 entries, 0 to 1048566
Data columns (total 9 columns):
 #   Column                   Non-Null Count    Dtype  
---  ------                   --------------    -----  
 0   TransactionID            1048567 non-null  object 
 1   CustomerID               1048567 non-null  object 
 2   CustomerDOB              1045170 non-null  object 
 3   CustGender               1047467 non-null  object 
 4   CustLocation             1048416 non-null  object 
 5   CustAccountBalance       1046198 non-null  float64
 6   TransactionDate          1048567 non-null  object 
 7   TransactionTime          1048567 non-null  int64  
 8   TransactionAmount (INR)  1048567 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 72.0+ MB
None
Index(['TransactionID', 'CustomerID', 'CustomerDOB', 'CustGender',
       'CustLocation', 'CustAccountBalance', 'TransactionDate',
       'TransactionTime', 'Trans

There are 1048567 rows and 9 columns


## Data Integrity

In [9]:
# Display rows with null values 
missing_rows = df[df.isnull().any(axis=1)] 
# print(missing_rows) 
# Number of rows with missing vales
print(len(missing_rows)) 
# Missing values count
null_values = df.isnull().sum()
print(null_values)
# Percentage of null values in each column
percent_null_values = null_values/nrows * 100
print('Percentage of null values: ', percent_null_values)


6953
TransactionID                 0
CustomerID                    0
CustomerDOB                3397
CustGender                 1100
CustLocation                151
CustAccountBalance         2369
TransactionDate               0
TransactionTime               0
TransactionAmount (INR)       0
dtype: int64
Percentage of null values:  TransactionID              0.000000
CustomerID                 0.000000
CustomerDOB                0.323966
CustGender                 0.104905
CustLocation               0.014401
CustAccountBalance         0.225927
TransactionDate            0.000000
TransactionTime            0.000000
TransactionAmount (INR)    0.000000
dtype: float64


- 6953 rows have missing values. This is abotu 0.663 % of the total number of rows
- TransactionID, CustomerID, TransactionDate, TransactionTime, TransactionAmount have no missing values.  
- CustomerDOB has 3397 values missing which is about  0.324 % of the number of values.
- CustGender has 1100 values missing which is about 0.104 % of the total number of values.
- CustLocation has 151 values missing which is about 0.0.014 % of the total number of values.
- CustAccountBalance has 2369 values missing which is about 0.226 % of the total number of values.

## Drop all missing rows

In [3]:
# Drop rows with missing values 
df=df.dropna(axis=0).reset_index(drop=True) 
# Verify rows are dropped 
print(df[df.isnull().any(axis=1)]) # verified 
# Update shape of dataframe 
print(df.shape)
print(df.info())
nrows, ncols = df.shape #  1041614 rows, 9 columns

Empty DataFrame
Columns: [TransactionID, CustomerID, CustomerDOB, CustGender, CustLocation, CustAccountBalance, TransactionDate, TransactionTime, TransactionAmount (INR)]
Index: []
(1041614, 9)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1041614 entries, 0 to 1041613
Data columns (total 9 columns):
 #   Column                   Non-Null Count    Dtype  
---  ------                   --------------    -----  
 0   TransactionID            1041614 non-null  object 
 1   CustomerID               1041614 non-null  object 
 2   CustomerDOB              1041614 non-null  object 
 3   CustGender               1041614 non-null  object 
 4   CustLocation             1041614 non-null  object 
 5   CustAccountBalance       1041614 non-null  float64
 6   TransactionDate          1041614 non-null  object 
 7   TransactionTime          1041614 non-null  int64  
 8   TransactionAmount (INR)  1041614 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 71.5+ MB
None


### Check for duplicated values

Only transaction Id needs to be checked. These cannot be duplicated since each transaction must be unique.

In [4]:
####---Check Redundant samples---#### 
print(df.nunique()) 
# No attributes with single value 
# ####---Check Duplicated Rows---#### 
# Calculate duplicates 
dupes = df.duplicated() # report if there are any duplicates 
print(dupes.any()) 
# List all duplicate rows 
print(df[dupes]) # duplicated rows


TransactionID              1041614
CustomerID                  879358
CustomerDOB                  17233
CustGender                       3
CustLocation                  9275
CustAccountBalance          160723
TransactionDate                 55
TransactionTime              81855
TransactionAmount (INR)      92391
dtype: int64
False
Empty DataFrame
Columns: [TransactionID, CustomerID, CustomerDOB, CustGender, CustLocation, CustAccountBalance, TransactionDate, TransactionTime, TransactionAmount (INR)]
Index: []


#### Save dataset


In [6]:
# Save cleaned dataframe to csv
df.to_csv(r'D:\Project\Bank customer segmentation\customer_segmentation\data\interim\bank_transctions_cleaned_I.csv')

### Advanced data integrity verfications


In [27]:
# Unique categories in data
# Customer gender
cust_gender_types = df['CustGender'].unique()    # 3 genders
# Customer location
cust_location = df['CustLocation'].unique()    # 9275 locations
# Customer ID
cust_id = df['CustomerID'].unique()    # 879358

### Advanced segmentation
- Segmentation by age  
- Segmentation by Transactions - amount, date, and time  
- Segmentation by account balance
- Segment by location aggregated to state or region

### Check Date of Birth

In [10]:
df['CustomerDOB'].value_counts()

1/1/1800    56292
1/1/89        809
1/1/90        784
6/8/91        698
1/1/91        665
            ...  
2/12/51         1
20/3/52         1
26/9/47         1
4/10/41         1
24/10/44        1
Name: CustomerDOB, Length: 17233, dtype: int64

## Save output

### Get unique customer ID